pandas/core/ops/array_ops.py

```diff
@@ -453,3 +442,87 @@ def get_array_op(op, str_rep: Optional[str] = None):
         return partial(logical_op, op=op)
     else:
         return partial(arithmetic_op, op=op, str_rep=str_rep)
+
+
+def maybe_upcast_datetimelike_array(obj: ArrayLike) -> ArrayLike:
+    """
+    If we have an ndarray that is either datetime64 or timedelta64, wrap in EA.
+
+    Parameters
+    ----------
+    obj : ndarray or ExtensionArray
+
+    Returns
+    -------
+    ndarray or ExtensionArray
+    """
+    if isinstance(obj, np.ndarray):
+        if obj.dtype.kind == "m":
+            from pandas.core.arrays import TimedeltaArray
+
+            return TimedeltaArray._from_sequence(obj)
+        if obj.dtype.kind == "M":
+            from pandas.core.arrays import DatetimeArray
+
+            return DatetimeArray._from_sequence(obj)
+
+    return obj
+
+
+def maybe_upcast_for_op(obj, shape: Tuple[int, ...]):
+    """
+    Cast non-pandas objects to pandas types to unify behavior of arithmetic
+    and comparison operations.
+
+    Parameters
+    ----------
+    obj: object
+    shape : tuple[int]
+
+    Returns
+    -------
+    out : object
+
+    Notes
+    -----
+    Be careful to call this *after* determining the `name` attribute to be
+    attached to the result of the arithmetic operation.
+    """
+    from pandas.core.arrays import DatetimeArray, TimedeltaArray
+
+    if type(obj) is timedelta:
```

Comment:
isinstance?

Reply:
we used type to exclude Timedelta; this func is moved unchanged from `__init__` since it is only used in this module